### DISCUSSION TITLE
SSH key verification with ManagedFile

### DISCUSSION THREAD

**Discussion by snerzkristall** [labgrid-project/labgrid#1007] [Q&A] [answered]
I am currently trying to flash a binary from a remote host. I use a custom driver (openocd driver with minor changes) for this which syncs a managedfile object to the machine running the exporter. Can i speficy ssh keys somewhere, because this is where my test breaks...?
Interacting via `labgrid-client console` is no problem it's just the ssh connection for syncing files.

**Current setup**
Windows 10 PC
             |
Docker container (labgrid client) --> needs ssh private key for syncing
             |
(local network)
             |
Raspberry Pi 4b (running exporter and coordinator) --> has ssh public key
             |
STM32 Boards (flash target)

Error log:
```
 transition state='start', args={'status': 'off'}
    off state='start'
    off state='stop', exception=ExecutionError(msg="failed to connect to <RASPI HOSTNAME> with args ['ssh', '-x', '-o', 'LogLevel=ERROR', '-o', 'PasswordAuthentication=no', '-n', '-MN', '-o', 'ConnectTimeout=30', '-o', 'ControlPersist=300', '-o', 'ControlMaster=yes', '-o', 'ControlPath=/tmp/labgrid-connection-drvj5yj0/control-<RASPI HOSTNAME>', '-o', 'StrictHostKeyChecking=yes',<RASPI HOSTNAME>], returncode=255 [],[b'No ECDSA host key is known for <RASPI HOSTNAME> and you have requested strict checking.\\r\\n', b'Host key verification failed.\\r\\n'] ", stdout=None, stderr=None), duration=0.121
```

**Comment by jluebbe** [comment, marked as answer]
For connecting to the exportes, labgrid just uses your user's ssh configuration. So if manual ssh <RASPI HOSTNAME> works there (in your container), it shouldn't be different for labgird.

**Reply by snerzkristall** [reply]
Thanks for the quick response, I managed to get ssh running, but there are still problems when connection via labgrid

**Example down below:**
Connecting to server-machine from docker container and accessing the SiSPM resource is not a problem.
```
(arnitestgrid-py3.10) root@20ea07000afb:~# ssh <RASPI-HOSTNAME>
Linux testbed 5.15.61-v8+ #1579 SMP PREEMPT Fri Aug 26 11:16:44 BST 2022 aarch64
...
license blabla
...
root@testbed:~# sispmctl -f 1
Accessing Gembird #0 USB device 011
Switched outlet 1 off
root@testbed:~#
```
However when doing this with `labgrid-client ...` it fails with some error, code 127.
```
(arnitestgrid-py3.10) root@20ea07000afb:~# labgrid-client -p heating power off
Traceback (most recent call last):
  File "/home/.cache/pypoetry/virtualenvs/arnitestgrid-QF4-4yUH-py3.10/lib/python3.10/site-packages/labgrid/remote/client.py", line 1821, in main
    args.func(session)
  File "/home/.cache/pypoetry/virtualenvs/arnitestgrid-QF4-4yUH-py3.10/lib/python3.10/site-packages/labgrid/remote/client.py", line 749, in power
    res = getattr(drv, action)()
  File "/home/.cache/pypoetry/virtualenvs/arnitestgrid-QF4-4yUH-py3.10/lib/python3.10/site-packages/labgrid/binding.py", line 96, in wrapper
    return func(self, *_args, **_kwargs)
  File "/home/.cache/pypoetry/virtualenvs/arnitestgrid-QF4-4yUH-py3.10/lib/python3.10/site-packages/labgrid/step.py", line 215, in wrapper
    _result = func(*_args, **_kwargs)
  File "/home/.cache/pypoetry/virtualenvs/arnitestgrid-QF4-4yUH-py3.10/lib/python3.10/site-packages/labgrid/driver/powerdriver.py", line 92, in off
    processwrapper.check_output(self._get_sispmctl_prefix() + cmd)
  File "/home/.cache/pypoetry/virtualenvs/arnitestgrid-QF4-4yUH-py3.10/lib/python3.10/site-packages/labgrid/step.py", line 215, in wrapper
    _result = func(*_args, **_kwargs)
  File "/home/.cache/pypoetry/virtualenvs/arnitestgrid-QF4-4yUH-py3.10/lib/python3.10/site-packages/labgrid/util/helper.py", line 95, in check_output
    raise subprocess.CalledProcessError(process.returncode,
subprocess.CalledProcessError: Command '['ssh', '-x', '-o', 'LogLevel=ERROR', '-o', 'PasswordAuthentication=no', '-o', 'ControlMaster=no', '-o', 'ControlPath=/tmp/labgrid-connection-kcu1ct3s/control-<RASPI-HOSTNAME>', '<RASPI-HOSTNAME>', '--', 'sispmctl', '-U', '001:011', '-f', '1']' returned non-zero exit status 127.
```

**Reply by snerzkristall** [reply]
**Update:**
Seems like it just can't find sispmctl, executing the exact same command outside of labgrid fails as well.
Both root and user pi on the server-machine have access to sispmctl (I edited /etc/profile to include the path/to/sispmctl and added the udev rules).
Did i forget something?

```
(arnitestgrid-py3.10) root@20ea07000afb:~/arnitestgrid# ssh -x -o LogLevel=ERROR -o PasswordAuthentication=no -o ControlMaster=no -o ControlPath=/tmp/labgrid-connection-knheqjt8/control-<RASPI-HOSTNAME> <RASPI-HOSTNAME> -- sispmctl -U 001:011 -f 1
bash: line 1: sispmctl: command not found
```

**Reply by Bastian-Krause** [reply]
Extending your PATH in `/etc/profile` will not affect SSH. Have a look at your `/etc/ssh/sshd_config` on your Raspberry Pi. There should be something like `# This sshd was compiled with PATH=/usr/bin:/bin:/usr/sbin:/sbin` in there. That's the PATH that is used.

What system is installed on the Pi? For Debian, there is a [sispmctl package](https://packages.debian.org/search?keywords=sispmctl).

**Reply by snerzkristall** [reply]
**Solution:**
ssh is using /usr/bin/local on the server-machine, if your executable is not there, the error above is thrown.

Create symbolic link `ln -s ` (not safe or reliable) or install on correct path to fix.

**Reply by snerzkristall** [reply]
I am using RaspbianOS (bullseye), the debian package (version 3.1) doesn't recognize my device...

**Reply by jluebbe** [reply]
sispmctl 4.1 was released in 2018, so updating the package in debian would be good, but won't help you in the short term. So perhaps just build and install it from source on your exporter.

**Reply by snerzkristall** [reply]
Ok thanks for the info.
I compiled 4.8 from source and it works without problems now.
```
(arnitestgrid-py3.10) root@20ea07000afb:~/arnitestgrid# labgrid-client -p heating power get
power for place heating is on
```

Thank you both a lot for the super quick and friendly responses :)
